KDP-5481 KDP-7511: ADD @Resource directive

NathanAtKadaster opened the pull request:
includes validation:
* only on IRI scalar output
* only once per GraphQL type
* cannot be nullable

NathanAtKadaster committed:
KDP-5481 KDP-7511: ADD @Resource directive

includes validation:
* only on IRI scalar output
* only once per GraphQL type
* cannot be nullable

JouriFledderman commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/ValueFetcher.java:
```diff
@@ -51,6 +52,11 @@ public Object get(DataFetchingEnvironment environment) {
     GraphQLType fieldType = GraphQLTypeUtil.unwrapNonNull(environment.getFieldType());
     QuerySolution source = environment.getSource();
 
+    if (Rdf4jScalars.IRI.getName()
```
Hier zou je volgens mij een check op aanwezigheid van @Resource directive moeten doen ipv op IRI type

NathanAtKadaster committed:
KDP-5481 KDP-7511: FIX ValueFetcher

Now checks if the Resource directive is present instead.
Checking if `type` == `IRI` doesn't do the trick right.

joostfarla commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/scalars/IriCoercing.java:
```diff
@@ -14,6 +14,10 @@
+      return (IRI) value;
+    }
+
     throw new UnsupportedOperationException();
```
Dit zou nu een ander soort exception moeten worden.

joostfarla commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/ConstructVerticeFactory.java:
```diff
@@ -35,7 +36,13 @@ public Vertice createRoot(@NonNull final Variable subject, @NonNull OuterQuery<?
       List<SelectedField> fields) {
-    List<Edge> edges = fields.stream()
+
+    List<SelectedField> filteredFields = fields.stream()
```
Dit zorgt er toch nog niet voor dat je op een IRI (subject) kan filteren (of sorteren)?

NathanAtKadaster replied:
nee dit zorgt ervoor dat de resource fields niet meegenomen worden in de verdere verwerking om vertices te creëren. De overige methodes vragen namelijk een propertyshape op die niet bestaat.

NathanAtKadaster committed:
KDP-5481 KDP-7511: FIX Sorting on a resource (in DB)

NathanAtKadaster committed:
KDP-5481 KDP-5859: FIX Filtering on a resource (in DB)

NathanAtKadaster committed:
KDP-5481 KDP-5859: REMOVE Prints

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/directives/ResourceDirectiveWiring.java:
```diff
@@ -0,0 +1,70 @@
+    return environment.getElement();
+  }
+
+  private void validateOnlyOnIri(String typename, String fieldname, GraphQLType rawType) {
```
Typename en fieldname worden alleen doorgegeven voor de Exceptie. Zou deze weglaten en de methode zelf een Exceptie laten gooien met een bericht. Maak de Exceptie een inner class Exception.

throw new ValidationException("can only be degined on a IRI field");

In de onFIeld methode deze afvangen (alle drie de validaties in een try block en dan in de catch een throw invalidConfigurationException("[GraphQL] Found an error on @{} directive defined on {}.{}: {} ", DIRECTIVE_NAME, typeName, fieldName, ex.getMessage());

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/directives/ResourceDirectiveWiring.java:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  private void validateOnlyOncePerType(String typeName, String fieldName, List<GraphQLFieldDefinition> fields) {
```
idem dito

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/directives/ResourceDirectiveWiring.java:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  private void validateOnlyRequired(String typeName, String fieldName, GraphQLFieldDefinition argument) {
```
idem dito

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/directives/ResourceDirectiveWiring.java:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  private void invalidConfigurationFor(String typeName, String fieldName, String reason) {
```
Weghalen en naar catch block onField methode.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/helper/CompareHelper.java:
```diff
@@ -26,15 +30,30 @@ private CompareHelper() {}
+      isResource = false;
+    }
+    return (value1, value2) -> compareOptional(resolveValue(value1, model, field, isResource, nodeShape),
+        resolveValue(value2, model, field, isResource, nodeShape), asc);
```
Dit is onleesbaar geworden. Idee om een getResourceComparator en een getOtherComparator (een goede omschrijvende naam) van te maken? Met aparte resolveResourceValue en resolveOtherValue.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/helper/CompareHelper.java:
```diff
@@ -26,15 +30,30 @@ private CompareHelper() {}
+      return Optional.of(value);
+    }
     String field = fields.remove(0);
+
```
Opsplitsen naar twee methodes om isResource argument weg te krijgen. De methode resolveResourceValue zou dan bestaan uit:

```
    if (fields.size() == 1) {
      return Optional.of(value);
    }
   return resolveOtherValue(...);
```

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/helper/CompareHelper.java:
```diff
@@ -48,8 +67,8 @@ private CompareHelper() {}
-        return childOptional
-            .flatMap(childValue -> resolveValue(childValue, model, String.join(".", field), propertyShape.getNode()));
+        return childOptional.flatMap(childValue -> resolveValue(childValue, model, String.join(".", field), isResource,
+            propertyShape.getNode()));
```
Blijft dit werken met bovenstaande suggestie? Oftewel is dieper de isResource nog nodig?

NathanAtKadaster replied:
ja die is nodig omdat je vanaf het pad steeds dieper gaat kijken. zolang er meer dan 1 waarde in het pad zit kun je nog een nodeshape/propertyshape ophalen dus vandaar ook die `fields#size`.

Wat denk ik wel werkt is als we die resolve Value niet recursief aanroepen en dat we van die `String path` een List maken.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -94,8 +96,23 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
+    List<DirectiveContainerObject> filterMapping =
+        coreTraverser.getTuples(environment, directiveWithValueFilter(CoreDirectives.FILTER_NAME))
+            .stream()
+            .peek(dco -> {
```
Geen gebruik maken van peek: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer- .

Doe dan eerst een collect en daarna de operatie op deze lijst (of visa versa).

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -94,8 +96,23 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
+
+              Optional.ofNullable(getField(environment.getFieldDefinition(), name.split("\\.")))
+                  .ifPresent(definition -> determineIsResource(dco, definition));
+
```
Idee om aparte methodes te maken met omschrijvende namen wat er functioneel gedaan wordt?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -114,13 +131,21 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
     return model.isEmpty() ? null : new QuerySolution(model, subjects.get(0));
   }
 
+  private void determineIsResource(DirectiveContainerObject directiveContainerObject,
```
De methode zegt dat je iets bepaalt, maar je verandert de staat van het directiveContainerObject naar setResource is false en ook naar true.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -161,7 +187,23 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
+        List<Map<String, String>> sortArgsMap = (List<Map<String, String>>) arguments.get(sortArgument.getName());
+
+        orderByObject = sortArgsMap.stream()
+            .peek(sortArgs -> {
```
Geen gebruik maken van peek. Zie ook comment boven.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -161,7 +187,23 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
+                    .getDirective(Rdf4jDirectives.RESOURCE_NAME));
+              }
+
+              resourceOptional.ifPresent(directive -> sortArgs.put("isResource", "true"));
```
Opsplitsen naar aparte functies met omschrijvende namen.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/QueryFetcher.java:
```diff
@@ -171,6 +213,16 @@ public Object get(@NonNull DataFetchingEnvironment environment) {
 
+  private GraphQLFieldDefinition getField(GraphQLFieldDefinition environment, String[] path) {
+    GraphQLFieldDefinition fieldDefinition =
+        ((GraphQLObjectType) GraphQLTypeUtil.unwrapAll(environment.getType())).getFieldDefinition(path[0]);
```
Kan na de if als return.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/ValueFetcher.java:
```diff
@@ -50,6 +51,11 @@ public ValueFetcher(final NodeShapeRegistry nodeShapeRegistry, Rdf4jConverterRou
     QuerySolution source = environment.getSource();
+    GraphQLFieldDefinition fieldDefinition = environment.getFieldDefinition();
+
+    if (Objects.nonNull(fieldDefinition.getDirective(Rdf4jDirectives.RESOURCE_NAME))) {
```
Eventueel methode maken isResource().

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -117,13 +123,34 @@ void processEdgeSort(Vertice vertice, GraphQLArgument argument, OuterQuery<?> qu
+          .getDirective(Rdf4jDirectives.RESOURCE_NAME));
+    }
+
+    resourceDirectiveOptional.ifPresent(directive -> orderMap.put("isResource", true));
```
orderMap.put("isResource", isResource()); en in de isResource methode de optional teruggeven.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -117,13 +123,34 @@ void processEdgeSort(Vertice vertice, GraphQLArgument argument, OuterQuery<?> qu
 
     String fieldName = orderMap.getOrDefault("field", argument.getName())
         .toString();
+    orderMap.put("isResource", false);
```
Is er een constante voor magic String isResource?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -117,13 +123,34 @@ void processEdgeSort(Vertice vertice, GraphQLArgument argument, OuterQuery<?> qu
 
-    if (Objects.nonNull(fieldName)) {
+    if (!(boolean) orderMap.get("isResource")) {
       findOrCreatePath(vertice, query, nodeShape.getPropertyShape(field.getName())
           .getNode(), new ArrayList<>(Arrays.asList(fieldName.split("\\."))), false, true);
```
Wat wordt hier functioneel gedaan?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -117,13 +123,34 @@ void processEdgeSort(Vertice vertice, GraphQLArgument argument, OuterQuery<?> qu
+    }
+    return fieldDefinition;
+  }
+
```
Deze methode ben ik eerder tegengekomen, zie boven idem dito verhaal.

NathanAtKadaster committed:
KDP-5481: IMPROVE validation exception handling

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -141,26 +168,35 @@ void processEdge(Vertice vertice, GraphQLArgument argument, OuterQuery<?> query,
+    } else {
+      Edge match = findOrCreatePath(vertice, query, nodeShape, filterRule.getPath(), true, false);
+      usedVertice = match.getObject();
+    }
```
Eventueel opsplitsen van stukken code naar aparte methodes om het leesbaarder te maken.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -141,26 +168,35 @@ void processEdge(Vertice vertice, GraphQLArgument argument, OuterQuery<?> query,
     Filter filter = createFilter(nodeShape, filterRule.getOperator(), filterRule.getValue(), filterRule.getPath()
         .get(filterRule.getPath()
-            .size() - 1));
+            .size() - 1),
```
Het is moeilijk te volgen wat hier gebeurd. Een idee om dit (getPath().size() -1) in de filterRule te zetten met een omschrijvende methode?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -141,26 +168,35 @@ void processEdge(Vertice vertice, GraphQLArgument argument, OuterQuery<?> query,
    * Create a new filter with either one argument or a list of arguments
    */
-  private Filter createFilter(NodeShape nodeShape, String filterOperator, Object filterValue, String argumentName) {
+  private Filter createFilter(NodeShape nodeShape, String filterOperator, Object filterValue, String argumentName,
```
Laatste vier argumenten vervangen door alleen FilterRule filterRule.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -169,9 +205,14 @@ private Filter createFilter(NodeShape nodeShape, String filterOperator, Object f
+          }
+          return getOperand(nodeShape, argumentName, serializerRouter.serialize(filterArgument),
+              rdf4jProperties.getShape()
+                  .getLanguage());
```
Aparte functie van maken voor leesbaarheid.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -169,9 +205,14 @@ private Filter createFilter(NodeShape nodeShape, String filterOperator, Object f
     }
```
Een aparte getFilterArguments(FilterValue filterValue) methode?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -262,14 +303,26 @@ void addOrderables(Vertice vertice, OuterQuery<?> query, Map<String, Object> ord
         .toString();
     String order = orderMap.get("order")
         .toString();
+    Boolean isResource = Boolean.parseBoolean((String) orderMap.get("isResource"));
```
isResource en ook andere magic Strings. Waarom is orderMap geen Klasse waar deze waarden uitgehaald kunnen worden?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/AbstractVerticeFactory.java:
```diff
@@ -262,14 +303,26 @@ void addOrderables(Vertice vertice, OuterQuery<?> query, Map<String, Object> ord
     Edge match;
     Variable subject;
-    if (nodeShape.equals(childShape)) {
+    if (isResource) {
```
Gehele functie addOrderables is erg lang nu dat er ook een isResource conditioneel deel in zit... opsplitsen naar meerdere methodes.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/ConstructVerticeFactory.java:
```diff
@@ -35,7 +36,13 @@ public Vertice createRoot(@NonNull final Variable subject, @NonNull OuterQuery<?
 
   private Vertice createVertice(final Variable subject, OuterQuery<?> query, NodeShape nodeShape,
```
Is een reeds bestaande functie, maar bestaat uit veel lines die de leesbaarheid niet ten goede komen.

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/SelectVerticeFactory.java:
```diff
@@ -40,46 +40,50 @@ public Vertice createVertice(Variable subject, OuterQuery<?> query, NodeShape no
+          Edge edge;
+
+          if (filter.getPath()
+              .size() == 1) {
```
Wat betekent .size() == 1 functioneel?

rregout commented on backend/rdf4j/src/main/java/org/dotwebstack/framework/backend/rdf4j/query/context/SelectVerticeFactory.java:
```diff
@@ -40,46 +40,50 @@ public Vertice createVertice(Variable subject, OuterQuery<?> query, NodeShape no
     Vertice vertice = createVertice(subject, nodeShape);
```
Ook een erg lange methode met allerlei condities.

rregout commented on backend/rdf4j/src/test/java/org/dotwebstack/framework/backend/rdf4j/ResourceIntegrationTest.java:
```diff
@@ -0,0 +1,154 @@
+
+  @Autowired
+  private GraphQL graphQL;
+
```
De drie testen hieronder lijken erg op elkaar. Valt volgens mij wel wat samen te voegen.

NathanAtKadaster committed:
KDP-5481: IMPROVE CompareHelper methods.

NathanAtKadaster committed:
KDP-5481: REMOVE usage of peek

NathanAtKadaster committed:
KDP-5481: REDUCE If statement complexity

INTRODUCE more functional style.
more consistency in used styles in a method.

msimons reviewed (changes requested):
Ik onderschrijf de opmerkingen van @rregout. Daarnaast moeten er nog checkstyle fouten weggewerkt worden. (spotless:apply)

NathanAtKadaster committed:
KDP-5481: REDUCE If statement complexity

INTRODUCE more functional style.
more consistency in used styles in a method.

NathanAtKadaster committed:
KDP-5481: REDUCE complexity VerticeFactories

NathanAtKadaster committed:
KDP-5481: CONDENSE ResourceIT using ArgumentsProvider

NathanAtKadaster commented:
Misschien iets om nog eens naar te kijken. Een `spotless:apply` werkt niet op `String`s. Die moet je eerst met de hand opknippen.

kad-bijste reviewed (changes requested):
Getest met aangepaste lokale versie.
Rest alleen nog het probleem bij de response op collecties met de MemIri datatype, waarbij een JSON serializatie fout optreed (met resource met simpleIri datatype ging de @resource wel goed).

NathanAtKadaster commented:
@bysterboz heb je hier een MVP voor wat je kan zippen? dan kan ik het lokaal ook produceren. Ik krijg het hier niet voor elkaar.

NathanAtKadaster commented:
fixed in #280
